File: core/liquidity/api/schemas.py
"""
Pydantic-схемы для запросов и ответов API.
"""
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Общие
# ---------------------------------------------------------------------------

class StatusResponse(BaseModel):
    status: str
    message: str | None = None


class ErrorResponse(BaseModel):
    detail: str


# ---------------------------------------------------------------------------
# ETL
# ---------------------------------------------------------------------------

class ETLRunRequest(BaseModel):
    source: Literal["postgres", "excel", "all"] = Field(
        "all", description="Источник данных для запуска ETL."
    )
    report_date: date | None = Field(
        None,
        description="Дата отчёта (по умолчанию сегодня).",
    )


class ETLRunResponse(BaseModel):
    batch_id:     int
    status:       str
    source:       str
    report_date:  date
    assets_loaded: int = 0
    liabs_loaded:  int = 0
    started_at:   datetime
    finished_at:  datetime | None = None
    error:        str | None = None


# ---------------------------------------------------------------------------
# Расчёты
# ---------------------------------------------------------------------------

class CalculateRequest(BaseModel):
    report_date: date = Field(..., description="Дата отчёта в формате YYYY-MM-DD.")
    calc_type: Literal["gap", "concentration", "full"] = Field(
        "full", description="Тип расчёта."
    )


class CalculationRecord(BaseModel):
    id:           int
    report_date:  date
    calc_type:    str
    status:       str
    started_at:   datetime
    finished_at:  datetime | None
    error_message: str | None

    class Config:
        from_attributes = True


class CalculateResponse(BaseModel):
    calculation_id: int
    status:         str
    report_date:    date
    calc_type:      str
    gap_rows:       int = 0
    conc_rows:      int = 0
    duration_sec:   float | None = None


# ---------------------------------------------------------------------------
# ГЭП-анализ
# ---------------------------------------------------------------------------

class GapBucketResult(BaseModel):
    bucket_code:           str
    bucket_name:           str
    sort_order:            int
    total_assets_rub:      float
    total_liabilities_rub: float
    gap_rub:               float
    cumulative_gap_rub:    float
    gap_ratio_pct:         float | None

    class Config:
        from_attributes = True


class GapAnalysisResponse(BaseModel):
    report_date:    date
    calculation_id: int
    buckets:        list[GapBucketResult]
    total_assets:   float
    total_liabilities: float
    net_gap:        float


# ---------------------------------------------------------------------------
# Концентрация
# ---------------------------------------------------------------------------

class ConcentrationItem(BaseModel):
    counterparty_code: str
    counterparty_name: str
    counterparty_type: str
    bucket_code:       str
    bucket_name:       str
    amount_rub:        float
    share_pct:         float

    class Config:
        from_attributes = True


class ConcentrationResponse(BaseModel):
    report_date:    date
    calculation_id: int
    category:       str
    items:          list[ConcentrationItem]
    total_amount:   float


# ---------------------------------------------------------------------------
# Отчёты
# ---------------------------------------------------------------------------

class ReportGenerateRequest(BaseModel):
    report_date:   date = Field(..., description="Дата отчёта.")
    report_type:   Literal["gap", "concentration", "full"] = "full"
    report_format: Literal["excel", "pdf", "csv"] = "excel"


class ReportTaskRecord(BaseModel):
    id:            int
    report_date:   date
    report_type:   str
    report_format: str
    report_name:   str | None
    status:        str
    file_path:     str | None
    error_message: str | None
    created_at:    datetime
    finished_at:   datetime | None

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Справочники
# ---------------------------------------------------------------------------

class TimebucketInfo(BaseModel):
    id:         int
    code:       str
    name:       str
    min_days:   int
    max_days:   int | None
    sort_order: int


class CounterpartyInfo(BaseModel):
    id:               int
    code:             str
    short_name:       str
    full_name:        str | None
    inn:              str | None
    counterparty_type: str | None
    country:          str | None

File: frontend/src/types/schemas.ts
export type CalcType = 'full' | 'gap' | 'concentration';

export interface CalculationRecord {
	id: number;
	report_date: string;
	calc_type: CalcType;
	status: string;
	started_at: string;
	finished_at: string | null;
	error_message: string | null;
}

export interface ConcentrationItem {
	counterparty_code: string;
	counterparty_name: string;
	counterparty_type: string;
	bucket_code: string;
	bucket_name: string;
	amount_rub: number;
	share_pct: number;
}

export interface GapBucketResult {
	bucket_code: string;
	bucket_name: string;
	sort_order: number;
	total_assets_rub: number;
	total_liabilities_rub: number;
	gap_rub: number;
	cumulitive_gap_rub: number;
	gap_ratio_pct: number;
}

export type ETLSource = 'all' | 'excel' | 'postgres';

export type ReportType = 'full' | 'gap' | 'concentration';
export type ReportFormat = 'excel' | 'csv' | 'pdf';

export interface ReportTaskRecord {
	id: number;
	report_date: string;
	report_type: ReportType;
	report_format: ReportFormat;
	report_name: string | null;
	status: string;
	file_path: string | null;
	error_message: string | null;
	created_at: string;
	finished_at: string | null;
}

export interface TimebucketInfo {
	id: number;
	code: string;
	name: string;
	min_days: number;
	max_days: number | null;
	sort_order: number;
};

export type CounterpartyType = 'bank' | 'cbr' | 'corporate' | 'individual';

export interface CounterpartyInfo {
  id: number;
  code: string;
  short_name: string;
  full_name: string;
  inn: string | null;
  counterparty_type: CounterpartyType;
  country: string;
}
